src/tsolvers/egraph/Enode.h

```diff
@@ -59,154 +82,86 @@ struct ELRef {
-//    explicit ELRef(uint32_t v) { x = v; }
 };
 
-// FIXME this is uninitialized right now.
 static struct ELRef ELRef_Undef = {INT32_MAX};
 
-class Extra {
-    struct {
-        ERef        car;
-        ERef        cdr;
-        ERef        next;           // Next node in the class
-        int         size;           // Size of the eq class
-        int         carParentIndex;
-        int         cdrParentIndex;
-    } lst;
-    struct {
-        PTRef       pterm;          // The corresponding pterm
-        ELRef       forbid;         // List of unmergeable Enodes
-        dist_t      dist_classes;   // The bit vector for distinction classes
-        int         dist_index;     // My distinction index
-        // fields related to explanation
-        PtAsgn      exp_reason;
-        ERef        exp_parent;
-        ERef        exp_root;
-        int         exp_time_stamp;
-    } trm;
-    friend class Enode;
-    friend class EnodeAllocator;
-};
-
 class EnodeAllocator;
 
-typedef uint32_t CgId;
-
-class Enode
+class Enode final
 {
 private:
-    // Defined for list and term Enodes
-    void setCar  (ERef er)       { assert(type() != et_symb); ex->lst.car = er; }
-    void setCdr  (ERef er)       { assert(type() != et_symb); ex->lst.cdr = er; }
-    void setPterm(PTRef tr)      { assert(isTerm()); ex->trm.pterm = tr; }
-
-protected:
     static uint32_t cgid_ctr;
 
-    struct {
-        unsigned type      : 2;
-        enodeid_t id       : 30; } header;
-
-    static_assert(sizeof(SymRef) == sizeof(ERef), "Expected size of types does not match");
-    union {
-        SymRef symb;
-        ERef   root;
-    };
-
-    ERef        er;             // Either my eref or reference to the relocated one
-    cgId        cid;            // The congruence id of the enode (defined also for symbols)
-    // Must be last field!
-    Extra       ex[0];          // Enode specific data
+    ERef    root;           // The root of this enode's equivalence class
+    cgId    cid;            // The congruence id of the enode (never changes)
+    ERef    eq_next;           // Next node in this enode's equivalence class
+    int     eq_size;           // Size of this enode's equivalence class
+    PTRef   pterm;          // The corresponding pterm
+    ELRef   forbid;         // List of unmergeable Enodes
+    dist_t  dist_classes;   // The bit vector for distinction classes
+
+    // fields related to explanation
+    PtAsgn      exp_reason;
+    ERef        exp_parent;
+    ERef        exp_root;
+    int         exp_time_stamp;
+
+    // Term representation
+    SymRef symb;
+    uint32_t argSize;
+
+    /*
+     * Pointer to the end of object, where ENodeAllocator allocated enough space for Enode's children and
+     * their use-vector indices.
+     */
+    ERef args[0];
 
     friend class EnodeAllocator;
-    friend class EnodeStore;
+    Enode(SymRef symbol, ERefSpan children, ERef myRef, PTRef ptr);
+    Enode(Enode const&) = delete;
```

Comment:
Apparently [deleted members should be public](https://stackoverflow.com/questions/18931133/must-a-deleted-constructor-be-private)

Reply:
Fixed.